lib/rubocop/cop/mavenlint/bigint_for_migration_keys.rb

```diff
@@ -0,0 +1,42 @@
+require 'rubocop'
+
+module RuboCop
+  module Cop
+    module Mavenlint
+      class BigIntForMigrationKeys < RuboCop::Cop::Cop
+        FK_VIOLATION_MSG = "Foreign keys must be of type BIGINT".freeze
+        PK_VIOLATION_MSG = "Primary keys must be of type BIGINT".freeze
+
+        def_node_matcher :change_column, <<~PATTERN
+          (send nil? :change_column ...)
+        PATTERN
+
+        def_node_matcher :add_column, <<~PATTERN
+          (send nil? :add_column ...)
+        PATTERN
+
+        def on_send(node)
+          return unless change_column(node) || add_column(node)
+
+          if is_foreign_id_column(node) && is_integer_column(node)
+            add_offense(node, message: FK_VIOLATION_MSG)
+          elsif is_primary_id_column(node) && is_integer_column(node)
+            add_offense(node, message: PK_VIOLATION_MSG)
+          end
+        end
+
+        def is_integer_column(node)
+          node.children[4].children.first == :integer
+        end
+
+        def is_foreign_id_column(node)
+          node.children[3].children.first.to_s.match(/.*_id/)
```

Comment:
Sorry if this is a basic linter question, but what does it mean to be a "child" of a node here? Is the node the method and the children the params?

Reply:
Rubocop linters are presented an AST (abstract syntax tree) that you can parse. For a migration the `up` node in the tree looks like this:

```
(send nil :change_column
  (sym :preferences)
  (sym :subject_id)
  (sym :bigint)
  (hash
    (pair
      (sym :unsigned)
      (false)
    )
    (pair
      (sym :null)
      (true)
    )
  )
)
```

The root `node` in this case is `send nil :change_column`. So `node.children[3]` is `(sym :bigint)`. You have to do `children.children` to get the actual value of the child node, not sure why the `ruby-parse` library works that way.

Reply:
Used this blog as an inspiration: https://downey.io/blog/writing-rubocop-linters-for-database-migrations/